### ISSUE TITLE
Faster addition of sparse diagonal matrices

### ISSUE THREAD

**Issue by nordam** [scipy/scipy#13984]
**Is your feature request related to a problem? Please describe.**
I'm writing a numerical code that makes use of ```scipy.sparse.dia_matrix```. For my application it is very natural and very readable to write code where I add together a few of these matrices, for example

```M = M1 + M2 + M3```

where all of these are diagonal sparse matrices (for my application, sometimes they are all tridiagonal, sometimes one has additional diagonals).

When profiling my code, I found that an inordinate amount of time was spent on these additions, and a user on Stackoverflow pointed out to me that the problem is that the matrices are converted to CSR format when you add them in this way. 
https://stackoverflow.com/questions/67388890/why-is-adding-instances-of-scipy-sparse-dia-matrix-so-slow

I am able to achieve far better performance by explicitly adding up the diagonals explicitly. For my tridiagonal matrices, this is an example:
```
M_ = dia_matrix((N, N))
for d in [-1, 0, 1]:
    M_.setdiag(M1.diagonal(d) + M2.diagonal(d) + M3.diagonal(d), d)
```
This works, but the code is far less readable and elegant. If I add them together directly, it looks basically identical to the equations I'm implementing, and clean code is after all one of the reasons I prefer Python.

**Describe the solution you'd like**

If at all practical, it would be nice if the add operation on instances of ```dia_matrix``` could check if the other matrix is also of type ```dia_matrix```, and if so, add the diagonals that are actually there, instead of converting to CSR.

**Comment by tupui** [comment]
So basically overwriting `__add__` in `dia_matrix` if I understand this correctly. Currently `__add__` this is inherited from `spmatrix`.

Since you managed to have it working, would you be willing to do a PR?

**Comment by nordam** [comment]
Sure, I can give it a try.

**Comment by nordam** [comment]
Ok, I've made a few attempts. Initially, I had a large number of failing tests, but now I'm I down to 6 failing tests. Unfortunately, I don't understand why they fail.

To give the complete background, here is the ```__add__``` method of ```scipy.sparse.spmatrix``` (from which ```dia_matrix``` inherits:

```
    def __add__(self, other):  # self + other
        if isscalarlike(other):
            if other == 0:
                return self.copy()
            # Now we would add this scalar to every element.
            raise NotImplementedError('adding a nonzero scalar to a '
                                      'sparse matrix is not supported')
        elif isspmatrix(other):
            if other.shape != self.shape:
                raise ValueError("inconsistent shapes")
            return self._add_sparse(other)
        elif isdense(other):
            other = np.broadcast_to(other, self.shape)
            return self._add_dense(other)
        else:
            return NotImplemented
```

Hence, what I think I need to to is only to write a custom ```_add_sparse``` for ```dia_matrix```. And again for background, here is the ```_add_sparse``` for ```spmatrix```:

```
    def _add_sparse(self, other):
        return self.tocsr()._add_sparse(other)
```

Here is my attempt at writing a new ```_add_sparse``` for ```dia_matrix```:

```
    def _add_sparse(self, other):
        """If other is also of type dia_matrix,
        add all the diagonals, and return a result of type dia_matrix.
        Otherwise, convert to csr_matrix, which is the default behaviour
        of parent class spmatrix.
        """

        # Check if other is also of type dia_matrix
        if not isinstance(other, type(self)):
            # If other is not of type dia_matrix, default to
            # converting to csr_matrix, as is done in the _add_sparse
            # method of parent class spmatrix
            return self.tocsr()._add_sparse(other)
        else:
            # The task is to compute m = self + other
            # Start by making an empty matrix, of the datatype
            # that should result from adding self and other
            m = dia_matrix(self.shape, dtype = np.result_type(self, other))
            # Add all the stored diagonals of self
            for d in self.offsets:
                m.setdiag(self.diagonal(d), d)
            # Then, add all the stored diagonals of other,
            # making sure to keep those we already added
            for d in other.offsets:
                m.setdiag(m.diagonal(d) + other.diagonal(d), d)
            # And return the result
            return m
```
As far as I can tell, the original ```__add__``` method of ```spmatrix``` is called first, and when it has confirmed that ```other``` is also a sparse matrix, and that the shapes are equal, it then calls ```_add_sparse```. So the shapes should always be equal when the function above is called.

The six failing tests are these:

```
======================= short test summary info =======================
FAILED scipy/sparse/tests/test_base.py::TestDIA::test_add_sub - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestDIA-test_add_sub] - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestDIA-test_add_sub] - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestDIA-test_add_sub] - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestDIA-test_add_sub] - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestDIA-test_add_sub] - ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
```

I can't really see why anything to do with shape should fail from the changes I've made. (And to be honest, I find it difficult to understand exactly what the resiliency tests are supposed to do, as they seem to be defined with some pytest features that I'm unfamiliar with.)

A point to note is that introducing this change _will_ change results at one point: Previously, adding together two instances of ```dia_matrix``` would give a ```csr_matrix```, but now (by design) it will return a ```dia_matrix``` instead. Maybe that could somehow throw a few of the tests?

**Comment by tupui** [comment]
I've never worked with these, but if that can help you here is the complete traceback for your issue:
```python
m.setdiag(m.diagonal(d) + other.diagonal(d), d)
Traceback (most recent call last):
  File ".../miniconda3/envs/scipydev/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3427, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-5a8e0c258460>", line 1, in <module>
    m.setdiag(m.diagonal(d) + other.diagonal(d), d)
  File ".../scipy/scipy/sparse/base.py", line 1162, in setdiag
    self._setdiag(np.asarray(values), k)
  File ".../scipy/scipy/sparse/dia.py", line 308, in _setdiag
    self.data[self.offsets == k, min_index:max_index] = values
ValueError: shape mismatch: value array of shape (3,)  could not be broadcast to indexing result of shape (1,1)
m.diagonal(d) + other.diagonal(d)
Out[9]: array([2.25 +0.j, 2.125+0.j, 2.25 +0.j])
d
Out[10]: 3
m.diagonal(d)
Out[11]: array([2.25+0.j])
other.diagonal(d)
Out[12]: array([ 0.   +0.j, -0.125+0.j,  0.   +0.j])
```

**Comment by nordam** [comment]
Ok, thanks! That's very helpful. I've been struggling to set up my environment so that I can run with the code I've written, but clearly I need to figure that out first.

**Comment by tupui** [comment]
My pleasure. Just in case: I am using PyCharm on Mac and it's very smooth.In this case I just added a breakpoint in the test and could pin-point rapidly the error.

**Comment by nordam** [comment]
I'm also on a mac, and I've installed and activated the environment with conda as described here: https://scipy.github.io/devdocs/dev/contributor/quickerstart_conda.html

runtests.py worked (at least the tests passed before I started making changes, and failed afterwards), and I also tried running ```python setup.py develop``` but when I try to run tests in a shell or a script, I get the default behaviour, as if I had made no changes.

**Comment by tupui** [comment]
> runtests.py worked (at least the tests passed before I started making changes, and failed afterwards), and I also tried running `python setup.py develop` but when I try to run tests in a shell or a script, I get the default behaviour, as if I had made no changes.

I think you are just missing `conda develop .`. We are still working on improving the developer documentation. But in your case this version is complete and should work: https://scipy.github.io/devdocs/dev/contributor/quickstart_mac.html#quickstart-mac

**Comment by nordam** [comment]
My apologies for going a bit off topic, but I tried setting up scipy development in PyCharm, and I can't get it to work. Upon opening the project, I said yes please when PyCharm asked it if should create a conda environment from the ```environment.yml``` file in the scipy repo. When all of that was done, I opened ```runtests.py```, and clicked the little green triangle in the top left corner of the editor. That kicked off building and running the tests, and seemed to work fine except that it didn't break at the break point I set.

Then I tried right-clicking ```runtests.py``` in the file browser panel, and selected "Debug 'runtests.py'", but that failed pretty quickly with complaints about being unable to find "math.h".

**Comment by tupui** [comment]
I would not use `runtests.py` here as you just want to launch a few tests (and I suppose it fails because PyCharm is launching this in a particular way). You should be able to right click on `test_base.py` and `Run...`. Or in the command line of PyCharm `pytest scipy/sparse/tests/test_base.py::TestDIA` and `pytest scipy/sparse/tests/test_base.py:: Test64Bit `

**Comment by nordam** [comment]
None of those work for me, I'm afraid: Clicking Run... on ```test_base.py``` gives this error:

```
.....
  File "/Users/username/anaconda3/envs/scipy-dev/lib/python3.9/warnings.py", line 262, in _getcategory
    raise _OptionError("invalid module name: %r" % (module,)) from None
warnings._OptionError: invalid module name: 'scipy._lib._testutils'

Process finished with exit code 1

Empty suite
```

and typing ```pytest scipy/sparse/tests/test_base.py::Test64Bit``` in the Terminal window of PyCharm gives me this error:

```
(base) ➜  scipy git:(master) ✗ pytest scipy/sparse/tests/test_base.py::Test64Bit
ImportError while loading conftest '/Users/username/Projects/scipy/scipy/scipy/conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('scipy.conftest', '/Users/username/anaconda3/envs/scipy-dev/lib/python3.9/site-packages/scipy/conftest.py', PosixPath('/Users/username/Projects/scipy/scipy/scipy/conftest.py'))
(base) ➜  scipy git:(master) ✗ 
```

**Comment by tupui** [comment]
Your PyCharm seem to be picking up your env. But the terminal does not seem ok as your are in `(base)`. 

Anyway, I think you need to build again. It's because when you are calling `runtests.py` it is building in a specific folder and then maybe your env does not find it.

What I usually do:

```bash
conda activate scipy-dev
python setup.py build_ext --inplace
conda develop .  # just needed once
```
Then in this env you should be able to do everything.

**Comment by nordam** [comment]
I gave up on PyCharm, but I got there in pdb in the end. The issue is the following:

During one of the tests, two instances of ```dia_matrix``` are added. One is of dtype float64, the other complex128 (not sure if the datatypes are relevant). Both the matrices have shape 4 x 6. Here are some relevant pieces of information from the debugger:

```
(Pdb) self.shape
(4, 6)
(Pdb) self.data.shape
(4, 4)
(Pdb) self.todense()
matrix([[-1.5  ,  6.5  ,  0.   ,  2.25 ,  0.   ,  0.   ],
        [ 3.125, -7.875,  0.625,  0.   ,  0.   ,  0.   ],
        [ 0.   ,  0.   , -0.125,  1.   ,  0.   ,  0.   ],
        [ 0.   ,  0.   ,  8.375,  0.   ,  0.   ,  0.   ]])
(Pdb) self.diagonal(3)
array([2.25])

(Pdb) other.shape
(4, 6)
(Pdb) other.data.shape
(7, 6)
(Pdb) other.todense()
matrix([[ 0.375 +1.25j  ,  0.    +0.j    ,  0.    +0.j    ,   0.    +0.j    , -5.    +6.j    ,  2.5   -3.875j ],
        [14.25  +2.25j  , -3.75  +4.125j ,  0.    +0.j    ,   0.    +0.j    , -0.125 +0.j    ,  0.    +2.75j  ],
        [ 0.    +0.j    ,  7.25  +4.125j ,  0.    +0.j    ,   0.    +0.j    ,  0.    +0.j    ,  0.    +0.j    ],
        [18.5   -0.0625j, -0.0625+0.j    ,  0.    +0.j    ,  0.    +0.j    ,  0.    +0.j    ,  0.    +0.j    ]])
(Pdb) other.diagonal(3)
array([ 0.   +0.j, -0.125+0.j,  0.   +0.j])
```

So the problem is that despite the fact that these are both 4 x 6 matrices, ```self.diagonal(3)``` returns ```array([2.25])```, instead of ```array([2.25, 0., 0.])```, which would be the correct 3rd superdiagonal (at least in normal mathematical terms). At the same time, ```other.diagonal(3)``` returns an array of the correct length: ```array([ 0.   +0.j, -0.125+0.j,  0.   +0.j])```. Hence, there is a shape mismatch when my code tries to add these diagonals together.

It might relevant here that the underlying array storing the data for ```self``` is of shape 4 x 4. At least, if I manually create a 4 x 6 instance of ```dia_matrix``` containing the same data as ```self```, then it behaves as I would expect, returning the full diagonal with the two zeros:

```
> m = diags([
    [3.125, 0., 8.375],
    [-1.5, -7.875, -0.125, 0.],
    [6.5, 0.625, 1, 0],
    [2.25, 0, 0]
], offsets = [-1, 0, 1, 3], shape = (4, 6))

> m.todense()
matrix([[-1.5  ,  6.5  ,  0.   ,  2.25 ,  0.   ,  0.   ],
        [ 3.125, -7.875,  0.625,  0.   ,  0.   ,  0.   ],
        [ 0.   ,  0.   , -0.125,  1.   ,  0.   ,  0.   ],
        [ 0.   ,  0.   ,  8.375,  0.   ,  0.   ,  0.   ]])

> m.diagonal(3)
array([2.25, 0.  , 0.  ])
```

 I have no idea why this happens. There is nothing in the documentation for ```dia_matrix.diagonal``` that indicates the diagonal might not be of full length.

**Comment by WarrenWeckesser** [comment]
> So the problem is that despite the fact that these are both 4 x 6 matrices, self.diagonal(3) returns array([2.25]), instead of array([2.25, 0., 0.])

That looks like a bug in the implementation of the `diagonal` method of the `dia_matrix` class.  Here's another example; `a.diagonal(0)` should return `array([1, 0])`, but it returns `array([1])`:

```
In [15]: a = dia_matrix([[1, 0, 0, 0], [0, 0, 0, 0]])

In [16]: a.A
Out[16]: 
array([[1, 0, 0, 0],
       [0, 0, 0, 0]])

In [17]: a.diagonal(0)
Out[17]: array([1])
```

**Comment by perimosocordiae** [comment]
Thanks for reporting, Warren! I have a fix ready for that issue in gh-14000. (An auspicious PR number, to be sure!)

@nordam if you'd like, you can paste in the few lines from my change to unblock your development.

**Comment by nordam** [comment]
I added the change from gh-14000 and then I got no failed tests in sparse.

**Comment by nordam** [comment]
I seet that gh-14000 has been merged into master now, so I guess if I merge master into my branch, then I can make a pull request?

How about unit tests? In principle, the only change that should result from my new ```_add_sparse``` function is that adding two instances of ```dia_matrix``` will now return a ```dia_matrix```, instead of a ```csr_matrix``` (and the addition should be faster). Should I add some tests to confirm that this works as intended?

**Comment by tupui** [comment]
Sure you can make a PR, thanks! As for the tests, sounds good but no need to check that it's faster IMO.

**Comment by perimosocordiae** [comment]
Feel free to open a PR with your change as is, then we can see if it needs more tests and look at some benchmarks.
